### PR TITLE
Simplify onResponse to remove unneeded promise layers.

### PR DESCRIPTION
**Pull request by yupinghu**
The wrapper's request function is already asynchronous, and
now that session handling is done earlier in middleware, we
can make requests a bit simpler. I also took the opportunity
to simplify how the response text is formed.

It should be possible to do the same sort of thing for
subscriptions but that will be a separate change.

### CODE REVIEW

File: lib/middleware/request-handler.ts [hunk 1]
```diff
@@ -190,30 +174,19 @@ export function onRequest(req: Interface.IOurRequest,
...
-                    }
-                });
-        }));
-    })()
-        .then(next)
-        .catch( (err: Error): any => {
-            req.log.error(err, 'Request error.');
-            return next(new restify.InternalError(err.message));
+    req.blpSession.request(util.format('//%s/%s', req.query.ns, req.query.service),
+        req.query.type,
```
Comment (ericvw) [inline lib/middleware/request-handler.ts]: nit: Align parameters

File: lib/middleware/request-handler.ts [hunk 2]
```diff
@@ -119,66 +119,50 @@ export function elevateRequest (req: Interface.IOurRequest,
...
-                chunk += ',';
-            }
-            chunk += stringifyPair( 'status', status ) + ',' +
-                     stringifyPair( 'message', message || '' ) +
-                     '}';
-            res.end( chunk );
-            needComma = false;
-            chunk = '';
-        });
+        var text: string = prepareResponseText(true);
```
Comment (ericvw) [inline lib/middleware/request-handler.ts]: nit: Type declaration not needed - can use inference.

File: lib/middleware/request-handler.ts [hunk 3]
```diff
@@ -119,66 +119,50 @@ export function elevateRequest (req: Interface.IOurRequest,
...
+                text = ',';
+            } else {
+                // This is the continuation of a response that has data and will not add more.
+                text = '],';
+            }
         }
-
-        return resultP || Promise.resolve(undefined);
-    };
+        return text;
```
Comment (ericvw) [inline lib/middleware/request-handler.ts:142]: We should assert that `text` is set.

File: lib/middleware/request-handler.ts [hunk 4]
```diff
@@ -119,66 +119,50 @@ export function elevateRequest (req: Interface.IOurRequest,
...
-    var chunk: string; // the current chunk string is assembled here
-    var needComma: boolean = false; // need to append a ',' before adding more data
 
-    // returns a promise
-    function prepareResponse (): Promise<any> {
-        var resultP: Promise<any> = undefined;
+    // This is called each time something
+    function prepareResponseText(isEnd: boolean): string {
+        var text: string;
         if (++chunkIndex === 1) {
```
Comment (ericvw) [inline lib/middleware/request-handler.ts]: nit: Reverse operands for the equality

File: lib/middleware/request-handler.ts [hunk 5]
```diff
@@ -119,66 +119,50 @@ export function elevateRequest (req: Interface.IOurRequest,
...
+            // This is the first time we're adding response text. Set the headers as well.
+            res.statusCode = 200;
             res.setHeader('content-type', 'application/json');
-            chunk = '{';
-            needComma = false;
+            if (!isEnd) {
```
Comment (ericlu88) [inline lib/middleware/request-handler.ts]: How about using conditional operator `?` here for conciseness?

File: lib/middleware/request-handler.ts [hunk 6]
```diff
@@ -119,66 +119,50 @@ export function elevateRequest (req: Interface.IOurRequest,
...
-        return resultP || Promise.resolve(undefined);
-    };
+        return text;
+    }
 
     function stringifyPair ( key: string, value: any ): string {
         return '"' + key.replace(/'/g, '\\$&') + '":' + JSON.stringify(value);
     }
 
     res.sendChunk = (data: any): Promise<void> => {
```
Comment (ericlu88) [inline lib/middleware/request-handler.ts:149]: It seems to me that there's no async operation in `sendChunk` and `sendEnd` method anymore. We should probably make it completely synchronously(ie, `res.sendChunk = (data: any): void`)

Comment (yupinghu) [inline lib/middleware/request-handler.ts:149]: Created #87 to track this -- I agree but subscriptions currently relies on this behavior and wasn't comfortable making that change all at once. I can tackle it after this one is in.

**Comment by ericvw** [comment]
Note: Let's not merge this until #83 is landed first.

**Comment by ericvw** [comment]
lgtm

**Comment by ericlu88** [comment]
LGTM too. Will hold off merging until #83 landed.

**Comment by ericvw** [comment]
@yupinghu - feel free to squash

**Pull request merged**